csharp/src/Ice/ObjectAdapter.cs

```diff
@@ -165,23 +205,133 @@ public async Task ActivateAsync(params DispatchInterceptor[] interceptors)
                 _interceptors.AddRange(Communicator.DispatchInterceptors);
                 _interceptors.AddRange(interceptors);
 
+                if (expandedEndpoints != null)
+                {
+                    Debug.Assert(_incomingConnectionFactories.Count == 0);
+
+                    _incomingConnectionFactories.AddRange(
+                        expandedEndpoints.Select<Endpoint, IncomingConnectionFactory>(
+                            endpoint =>
+                            endpoint.IsDatagram ?
+                                new DatagramIncomingConnectionFactory(this, endpoint) :
+                                new AcceptorIncomingConnectionFactory(this, endpoint)));
+                }
+
                 // Activate the incoming connection factories to start accepting connections
                 foreach (IncomingConnectionFactory factory in _incomingConnectionFactories)
                 {
                     factory.Activate();
                 }
 
-                // In the event _publishedEndpoints is empty, RegisterEndpointsAsync does nothing.
-                _activateTask = RegisterEndpointsAsync(PublishedEndpoints, default);
+                _activateTask = _routerInfo != null ? ActivateWithRouterAsync(cancel) : RegisterEndpointsAsync(cancel);
```

Comment:
Is this correct, we don't register the endpoints for router adapters?

Reply:
Yes: we actually don't read the endpoints for a routed adapters - they are endpoint-less.